### PR TITLE
Bricksort parallel implemented

### LINKED ISSUE #129: Parallel Brick Sort
#### Description of the problem
I would like to propose an odd-even transposition sort or brick sort. Odd-Even Transposition Sort is based on the Bubble Sort technique. It compares two adjacent numbers and switches them if the first number is greater than the second number to get an ascending order list. The opposite case applies to a descending order series.

Please assign me this issue.

### PR DESCRIPTION
**Pull request by HarsheetKakar**
#### References to other Issues or PRs or Relevant literature

Fixes #129 

#### Brief description of what is fixed or changed
Thread based bricksort implemented

### CODE REVIEW

File: pydatastructs/linear_data_structures/algorithms.py [hunk 1]
```diff
@@ -163,3 +164,69 @@ def brick_sort(array, **kwargs):
...
+    comp = kwargs.get("comp", lambda u, v: u <= v)
+
+    is_sorted = False
+    while is_sorted is False:
+        is_sorted = True
+        with ThreadPoolExecutor(max_workers=2) as executor:
+            t1 = executor.submit(_brick_sort_parallel_helper, array, start, end, comp, 1)
+            t2 = executor.submit(_brick_sort_parallel_helper, array, start, end, comp, 0)
+
+            is_sorted = t1.result() and t2.result()
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]: A better way to parallelize brick sort is two compare and swap elements on multiple threads. The current implementation leads too many races(which may be benign but reduce the efficiency).
See, [this image](https://www.google.com/url?sa=i&url=http%3A%2F%2Fwww.darshan.ac.in%2FUpload%2FDIET%2FDocuments%2FCE%2F180702_PP%2520-%2520IMP%2520Questions%2520and%2520Solution%2520-%25202015_17042015_061013AM.pdf&psig=AOvVaw1jFpyWZvus_IqgOe19ZdSh&ust=1585406147731000&source=images&cd=vfe&ved=0CAIQjRxqFwoTCNjco63wuugCFQAAAAAdAAAAABAI) to understand my point.

Comment (HarsheetKakar) [reply pydatastructs/linear_data_structures/algorithms.py]: the function which has i = 1 will work on odd indexes and i = 0 will work on even indexes. The link is sending me to a pdf, shall I read it whole or is there a specific page number you want me to refer?

Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py]: download
See here, various pair of indices can be parallely compared independently without any race in a specific phase. So, after running even number of phases we can check through `is_sort` whether the array is sorted or not.

Comment (HarsheetKakar) [reply pydatastructs/linear_data_structures/algorithms.py]: ok correct me if I am wrong, you are saying that to assign each swap in Phase 1 its own thread which compares and swaps the their 2 indexes and when phase 1 is completed then phase 2 should start. So concurrency in the Phase itself but not between the 2 phases?

Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py]: Yes, like if we are using some `num_threads` number of threads then the swaps can be distributed among those threads. Though the races are benign if two threads start from different indices but having two threads instead of one isn't really that useful. The time will be reduced by at most half.

Comment (HarsheetKakar) [reply pydatastructs/linear_data_structures/algorithms.py]: I really like the idea, will make the required changes by tomorrow.

Comment (HarsheetKakar) [reply pydatastructs/linear_data_structures/algorithms.py]: I have made the required changes, please check it out

File: pydatastructs/linear_data_structures/algorithms.py [hunk 2]
```diff
@@ -163,3 +164,75 @@ def brick_sort(array, **kwargs):
 
     if _check_type(array, DynamicArray):
         array._modify(force=True)
+
+def _brick_sort_comp(array, i, j, comp):
+    if(_comp(array[j], array[i], comp)):
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]:
```suggestion
    if _comp(array[j], array[i], comp):
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 3]
```diff
@@ -163,3 +164,75 @@ def brick_sort(array, **kwargs):
 
     if _check_type(array, DynamicArray):
         array._modify(force=True)
+
+def _brick_sort_comp(array, i, j, comp):
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]: Why is this needed?

Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]:
```suggestion
def _brick_sort_swap(array, i, j, comp):
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 4]
```diff
@@ -163,3 +164,75 @@ def brick_sort(array, **kwargs):
...
+    end = kwargs.get('end', len(array) - 1)
+    comp = kwargs.get("comp", lambda u, v: u <= v)
+
+    is_sorted = False
+    with ThreadPoolExecutor(max_workers=num_threads) as executor:
+        while is_sorted is False:
+            is_sorted = True
+            futures = []
+            for i in range(start+1, end, 2):
+                futures.append(executor.submit(_brick_sort_comp, array, i, i+1, comp))
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]:
```suggestion
                futures.append(executor.submit(_brick_sort_swap, array, i, i+1, comp))
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 5]
```diff
@@ -163,3 +164,75 @@ def brick_sort(array, **kwargs):
...
+            is_sorted = all(i.result() for i in futures)
+
+            futures = []
+            for i in range(start, end, 2):
+                futures.append(executor.submit(_brick_sort_comp, array, i, i+1, comp))
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]:
```suggestion
                futures.append(executor.submit(_brick_sort_comp, array, i, i + 1, comp))
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 6]
```diff
@@ -163,3 +164,75 @@ def brick_sort(array, **kwargs):
 
     if _check_type(array, DynamicArray):
         array._modify(force=True)
+
+def _brick_sort_comp(array, i, j, comp):
+    if(_comp(array[j], array[i], comp)):
+        array[i], array[j] = array[j], array[i]
+        return False
+    return True
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]:
```suggestion
def _brick_sort_comp(array, i, j, comp, is_sorted):
    is_sorted = _comp(array[i], array[j], comp)
    if not is_sorted:
        array[i], array[j] = array[j], array[i]
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 7]
```diff
@@ -163,3 +164,75 @@ def brick_sort(array, **kwargs):
...
+                futures.append(executor.submit(_brick_sort_comp, array, i, i+1, comp))
+            is_sorted = all(i.result() for i in futures)
+
+            futures = []
+            for i in range(start, end, 2):
+                futures.append(executor.submit(_brick_sort_comp, array, i, i+1, comp))
+            is_sorted = all(i.result() for i in futures)
+
+    if _check_type(array, DynamicArray):
+        array._modify(force=True)
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]:
```suggestion
    start = kwargs.get('start', 0)
    end = kwargs.get('end', len(array) - 1)
    comp = kwargs.get('comp', lambda u, v: u <= v)

    is_sorted = [False]
    with ThreadPoolExecutor(max_workers=num_threads) as executor:
        while is_sorted[0] is False:
            is_sorted[0] = True
            for i in range(start + 1, end, 2):
                executor.submit(_brick_sort_comp, array, i, i+1, comp, is_sorted).result()

            for i in range(start, end, 2):
                futures.append(executor.submit(_brick_sort_comp, array, i, i+1, comp, is_sorted))

    if _check_type(array, DynamicArray):
        array._modify(force=True)
```

File: pydatastructs/linear_data_structures/algorithms.py (hunk 6 above)
Comment (HarsheetKakar) [reply pydatastructs/linear_data_structures/algorithms.py]: wouldnt this cause race condition since all the threads will try to access is_sorted

**Comment by czgdp1807** [comment]
I have made some suggestions. Using `futures` isn't useful,

1. Consumes extra memory.
2. The check using `all` is sequential which makes doing comparisons using threads useless. 

The suggestions pass `is_sorted` as a list(for passing by reference automatically) and the first element is made `False` whenever a thread swaps and doesn't change if no swap occurs or in simple words, write to `is_sorted[0]` happens only when a swap occurs.
It should work but I am not very sure that it will.

**Comment by HarsheetKakar** [comment]
> I have made some suggestions. Using `futures` isn't useful,
> 
> 1. Consumes extra memory.
> 2. The check using `all` is sequential which makes doing comparisons using threads useless.
> 
> The suggestions pass `is_sorted` as a list(for passing by reference automatically) and the first element is made `False` whenever a thread swaps and doesn't change if no swap occurs or in simple words, write to `is_sorted[0]` happens only when a swap occurs.
> It should work but I am not very sure that it will.

I get it, working on it now

File: pydatastructs/linear_data_structures/algorithms.py (hunk 6 above)
Comment (czgdp1807) [reply pydatastructs/linear_data_structures/algorithms.py]: Well, we want to run the next pass only when at least a single swap occurs in the current pass. This is what it represents.

Comment (HarsheetKakar) [reply pydatastructs/linear_data_structures/algorithms.py]: got it. I thought that `result()` works like `join()` and will wait till the function returns a result.

File: pydatastructs/linear_data_structures/algorithms.py [hunk 8]
```diff
@@ -163,3 +164,70 @@ def brick_sort(array, **kwargs):
 
     if _check_type(array, DynamicArray):
         array._modify(force=True)
+
+def _brick_sort_swap(array, i, j, comp, is_sorted):
+    if(_comp(array[j], array[i], comp)):
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]:
```suggestion
    if _comp(array[j], array[i], comp):
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 9]
```diff
@@ -163,3 +164,70 @@ def brick_sort(array, **kwargs):
...
+        is_sorted[0] = False
+
+def brick_sort_parallel(array, num_threads, **kwargs):
+    """
+    Implements Concurrent Brick Sort / Odd Even sorting algorithm
+
+    Parameters
+    ==========
+
+    array: Array
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]:
```suggestion
    array: Array/list
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 10]
```diff
@@ -163,3 +164,70 @@ def brick_sort(array, **kwargs):
...
+    >>> arr = OneDimensionalArray(int,[3, 2, 1])
+    >>> brick_sort_parallel(arr, num_threads=5)
+    >>> [arr[0], arr[1], arr[2]]
+    [1, 2, 3]
+    >>> brick_sort_parallel(arr, num_threads=5, comp=lambda u, v: u > v)
+    >>> [arr[0], arr[1], arr[2]]
+    [3, 2, 1]
+
+    References
+    ==========
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]:
```suggestion
    ==========

```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 11]
```diff
@@ -163,3 +164,70 @@ def brick_sort(array, **kwargs):
...
+    .. [1] https://www.geeksforgeeks.org/odd-even-sort-brick-sort/
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]:
```suggestion
    .. [1] https://en.wikipedia.org/wiki/Odd%E2%80%93even_sort
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 12]
```diff
@@ -163,3 +164,70 @@ def brick_sort(array, **kwargs):
...
+
+    start = kwargs.get('start', 0)
+    end = kwargs.get('end', len(array) - 1)
+    comp = kwargs.get("comp", lambda u, v: u <= v)
+
+    is_sorted = [False]
+    with ThreadPoolExecutor(max_workers=num_threads) as executor:
+        while is_sorted[0] is False:
+            is_sorted[0] = True
+            for i in range(start+1, end, 2):
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]:
```suggestion
            for i in range(start + 1, end, 2):
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 13]
```diff
@@ -163,3 +164,71 @@ def brick_sort(array, **kwargs):
...
+
+    is_sorted = [False]
+    with ThreadPoolExecutor(max_workers=num_threads) as executor:
+        while is_sorted[0] is False:
+            is_sorted[0] = True
+            for i in range(start + 1, end, 2):
+                executor.submit(_brick_sort_swap, array, i, i + 1, comp, is_sorted).result()
+
+            for i in range(start, end, 2):
+                executor.submit(_brick_sort_swap, array, i, i + 1, comp, is_sorted).result()
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]:
```suggestion
    with ThreadPoolExecutor(max_workers=num_threads) as Executor:
        while is_sorted[0] is False:
            is_sorted[0] = True
            for i in range(start + 1, end, 2):
                Executor.submit(_brick_sort_swap, array, i, i + 1, comp, is_sorted).result()

            for i in range(start, end, 2):
                Executor.submit(_brick_sort_swap, array, i, i + 1, comp, is_sorted).result()
```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 14]
```diff
@@ -163,3 +164,71 @@ def brick_sort(array, **kwargs):
...
+    comp: lambda/function
+        The comparator which is to be used
+        for sorting. If the function returns
+        False then only swapping is performed.
+        Optional, by default, less than or
+        equal to is used for comparing two
+        values.
+
+    Examples
+    ========
```
Comment (czgdp1807) [inline pydatastructs/linear_data_structures/algorithms.py]:
```suggestion
    ========

```

File: pydatastructs/linear_data_structures/algorithms.py [hunk 15]
```diff
@@ -166,7 +166,7 @@ def brick_sort(array, **kwargs):
         array._modify(force=True)
 
 def _brick_sort_swap(array, i, j, comp, is_sorted):
-    if(_comp(array[j], array[i], comp)):
+    if _comp(array[j], array[i], comp):
```
Comment (HarsheetKakar) [inline pydatastructs/linear_data_structures/algorithms.py]: sorry forgot about this one

**Comment by HarsheetKakar** [comment]
good to go? @czgdp1807
